### PR TITLE
Implement `#[macro_export(local_inner_macros)]` (a solution for the macro helper import problem)

### PR DESCRIPTION
**Pull request by petrochenkov**
Implement a solution for the macro helper issue discussed in https://github.com/rust-lang/rust/issues/35896 as described in https://github.com/rust-lang/rust/issues/35896#issuecomment-395977901.

Macros exported from libraries can be marked with `#[macro_export(local_inner_macros)]` and this annotation changes how nested macros in them are resolved.

If we have a fn-like macro call `ident!(...)` and `ident` comes from a `macro_rules!` macro marked with  `#[macro_export(local_inner_macros)]` then it's replaced with `$crate::ident!(...)` and resolved as such (`$crate` gets the same context as `ident`).

### CODE REVIEW

**Comment by petrochenkov** [comment]
r? @nikomatsakis 

cc @dtolnay @rust-lang/lang 
I've added only a basic test, if you are interested in some test cases, please tell! I'll add them as well.

**Comment by petrochenkov** [comment]
There's one subtle detail that affects how we treat the [`nom`-like](https://github.com/rust-lang/rust/issues/35896#issuecomment-395599739) cases.

If we have a macro call using a macro variable, like `$var!(...)` where `$var` ultimately means `ident`, then we check whether that `ident` comes from a `macro_helper_hack` macro, not the macro call `$var!(...)` itself.
```rust
// Library crate
#[macro_export(macro_helper_hack)]
macro_rules! public {
    ($helper: ident) => {
        $helper!();
    }
}

// User crate
// `my_helper` is not affected by the macro helper hack (not changed to `$crate::my_helper`) and is resolved in the user crate, not library crate
public!(my_helper);
```

**Comment by dtolnay** [comment]
Currently building locally to play around with this, but I think this does the trick. Grabbing the span from the ident is the right behavior (as opposed to from the macro call).

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -1517,6 +1520,16 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
             }
         }
 
+        if attr.check_name("macro_export") {
+            if let Some(content) = attr.meta_item_list() {
+                if attr::list_contains_name(&content, "macro_helper_hack") {
+                    gate_feature_post!(&self, macro_helper_hack, attr.span,
```
Comment (petrochenkov) [inline src/libsyntax/feature_gate.rs]: Well, this must be a mistake, the attribute shouldn't be gated at the library side, but rather the effect should be enabled if `feature(use_extern_macros)` is enabled on the use site or stabilized.

Comment (dtolnay) [reply src/libsyntax/feature_gate.rs]: I was about to write the same thing. This should never happen:

```
error[E0658]: `macro_helper_hack` in `macro_export` is experimental (see issue #35896)
  --> serde-json/src/macros.rs:69:1
   |
69 | #[macro_export(macro_helper_hack)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(macro_helper_hack)] to the crate attributes to enable
```

If not building with `use_extern_macros`, the `macro_helper_hack` should have no effect and should be ignored as it has been on all past stable compilers.

**Comment by dtolnay** [comment]
Overall I found that this PR works incredibly well. I tested it against some of my nastiest macro code and it behaved as desired every time, with nom-style passing of macro names across crates and everything.

The one exception I found was it seems `macro_export(macro_helper_hack)` macros cannot be invoked from inside of their own crate:

```rust
#![feature(use_extern_macros, macro_helper_hack)]

#[macro_export(macro_helper_hack)]
macro_rules! noop {
    () => {
        noop_helper!{}
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! noop_helper {
    () => {};
}

fn main() {
    noop!{}
}
```

```
error[E0433]: failed to resolve. Maybe a missing `extern crate noop_helper;`?
  --> src/main.rs:6:9
   |
6  |         noop_helper!{}
   |         ^^^^^^^^^^^ Maybe a missing `extern crate noop_helper;`?
...
17 |     noop!{}
   |     ------- in this macro invocation
```

**Comment by dtolnay** [comment]
Ah, it appears the problem is not with this PR. The same thing happens invoking a macro through $crate just with `use_extern_macros`.

```rust
#![feature(use_extern_macros)]

#[macro_export]
macro_rules! noop {
    () => {
        $crate::noop_helper!{}
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! noop_helper {
    () => {};
}

pub fn call() {
    noop!{}
}
```

```
error[E0433]: failed to resolve. Maybe a missing `extern crate noop_helper;`?
  --> src/lib.rs:6:17
   |
6  |         $crate::noop_helper!{}
   |                 ^^^^^^^^^^^ Maybe a missing `extern crate noop_helper;`?
...
17 |     noop!{}
   |     ------- in this macro invocation
```

**Comment by dtolnay** [comment]
I guess this is a consequence of https://github.com/rust-lang/rust/issues/35896#issuecomment-378271859.

> macro_rules! macros declared within the same crate still use the legacy visibility rules, and cannot be imported using use statements and path syntax

Seems easy to work around in any case, and uncommon anyway.

**Comment by petrochenkov** [comment]
Yes, `$crate::foo!()` where `foo` is local `macro_rules!` doesn't work if called in the same crate, this is the primary drawback of both this PR and explicitly writing `$crate::foo!()`.

With [duplicated `macro_export`s](https://github.com/rust-lang/rust/pull/50143) prohibited this should be fixable if `$crate::foo!()` is defined in the same way as it's visible from other crates, i.e. referring to the corresponding unique `macro_export`ed macro and not to one of the macros named `foo` and defined in the root module.
```rust
macro_rules! foo { () => () } // (1)

mod m {
    // maybe somewhere in deeply nested modules
    #[macro_export] macro_rules! foo { () => () } (2)
}

macro_rules! foo { () => () } // (3), shadows (1) for the rest of the module

$crate::foo!(); // refers to (2), like from other crates, not to (1) or (3)
```

This doesn't seem urgent and blocking `use_extern_macros` though, because it's mostly about library's implementation conveniences and not about public interface.

**Comment by dtolnay** [comment]
Once ```"`macro_helper_hack` in `macro_export` is experimental"``` no longer happens (or if the merge is timed to never end up feature gated in a stable release), this resolves my concern about use_extern_macros. :+1:

**Comment by petrochenkov** [comment]
Updated.
- Feature gate is removed.
- `macro_helper_hack` is renamed into `local_inner_macros`

Bikeshedding about the attribute name is welcome.
`local_inner_macros` seemed like a reasonable name to me, but perhaps we should keep `macro_helper_hack` instead? :)
Any other suggestions?

**Comment by joshtriplett** [comment]
:+1: for `local_inner_macros`.

**Comment by nikomatsakis** [comment]
@petrochenkov sorry for being slow to provide feedback. Mozilla All Hands was all consuming. I'll take a look and also skim the comments, but before that I wanted to say thanks for trying this out!

**Comment by nikomatsakis** [comment]
Well, I'm happy that @dtolnay is happy. It seems like it'd be good to get feedback from @nrc, who was one of the people pushing against "mixing the systems".

cc @rust-lang/lang

**Comment by petrochenkov** [comment]
Let's notify people formally, there's a number of other things that are blocked on this PR.
Also marking as nominated to increase visibility.

@rfcbot fcp merge

This PR implements a solution for the macro helper issue described by dtolnay in https://github.com/rust-lang/rust/issues/35896#issuecomment-390381120.

Macros exported from libraries can be marked with `#[macro_export(local_inner_macros)]` and this annotation changes how nested macros in them are resolved.

If we have a fn-like macro call `ident!(...)` and `ident` comes from a `macro_rules!` macro marked with  `#[macro_export(local_inner_macros)]` then it's replaced with `$crate::ident!(...)` and resolved as such (`$crate` gets the same context as `ident`).

**Comment by petrochenkov** [comment]
@rfcbot still doesn't listen to me

**Comment by withoutboats** [comment]
@rfcbot concern testing-that-it-works

This PR clearly works for some macros (in particular, @dtolnay's), but its not clear to me if it will work for making all macro_rules macros importable, or if it doesn't, how many macros fall into the space where they'd have to bump their minimum version instead of using this feature. And I don't know an easy way to get a sense of that, either. Its not like we can do a crater run.

**Comment by dtolnay** [comment]
@withoutboats I believe this will work for practically all macro_rules macros. The exception is if one macro refers to *both* some private helpers and some other macros that are the caller's responsibility to have in scope.

#### Library's lib.rs

```rust
#[macro_export]
macro_rules! call_my_callback {
    () => {
        // Problematic because tagging local_inner_macros would look for
        // callback in this crate.
        call_my_callback_helper! { callback }
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! call_my_callback_helper {
    ($callback:ident) => {
        $callback!{}
    };
}
```

#### User's main.rs

```rust
macro_rules! callback {
    () => {
        println!("called.");
    };
}

fn main() {
    call_my_callback!();
}
```

In such cases the thing to do would be split that macro using both local helpers and caller's macros into two macros, a local_inner_macros entry point that passes the names of all local helpers to the next step, and a non-local_inner_macros helper that uses the helpers it receives along with the macros from the caller. I believe splitting a macro this way is always possible and straightforward. Even so I don't know of any macros where this would need to be applied.

```rust
#[macro_export(local_inner_macros)]
macro_rules! call_my_callback {
    () => {
        call_my_callback_with_helpers! { call_my_callback_helper }
    };
}

#[macro_export]
macro_rules! call_my_callback_with_helpers {
    ($helper:ident) => {
        $helper! { callback }
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! call_my_callback_helper {
    ($callback:ident) => {
        $callback!{}
    };
}
```

**Comment by withoutboats** [comment]
It sounds like:

1. there's no case where a macro passed explicitly as input could change its resolution based on this PR?
2. any time you're relying on dynamic resolution and also local helpers, you should be able to split your macro into two?

If both of these are true, it seems like we have all of our bases covered.

**Comment by withoutboats** [comment]
I'm also concerned about the long term. For people who don't care about backward compatibility to previous versions, we'd prefer them to use `$crate::` explicitly, right? How can we encourage that behavior in greenfield code?

**Comment by nikomatsakis** [comment]
I think this doesn't have to work 100%, it has to work for the crates we care about.

In terms of pushing people away, I think we can always document that this is a "backwards compatibility kludge" and is not the recommended way to do things (which is to use `$crate`). If people *do* wind up using it, oh well? We have to support it either way.

**Comment by dtolnay** [comment]
> 1. there's no case where a macro passed explicitly as input could change its resolution based on this PR?
> 2. any time you're relying on dynamic resolution and also local helpers, you should be able to split your macro into two?

I believe both of these are true.

**Comment by petrochenkov** [comment]
>so at least by that point people will necessarily use $crate.

`macro` macros don't even need `$crate`, they automatically resolve absolute paths at def-site.

**Comment by withoutboats** [comment]
@rfcbot resolve testing-that-it-works

**Comment by nikomatsakis** [comment]
@rfcbot concern what-are-we-fcping

It's not very clear to me, to be honest, just what we are "FCPing" here. My assumption was that we were FCP'ing a new, unstable attribute (presumably feature-gated). We would then separately have a stabilization discussion of `use_extern_macros` in conjunction with this feature. Is that accurate?

**Comment by joshtriplett** [comment]
@nikomatsakis I had the same assumption, and checked my box on the basis that that was what I was signing off on. I certainly hope that assumption matches what we're actually doing here.

**Comment by dtolnay** [comment]
@nikomatsakis @joshtriplett this PR only applies if `feature(use_extern_macros)` is set. The only line of code in the PR that affects behavior is surrounded by:

```rust
if self.use_extern_macros && ...
```

Without `use_extern_macros` a `#[macro_export(local_inner_macros)]` line will continue to be treated exactly like a plain `#[macro_export]` as it has been in every previous Rust release.

**Comment by cramertj** [comment]
@nikomatsakis @joshtriplett No, we can't feature-gate the attribute because then it would be an error to use this syntax in previously-published versions of rustc.

**Comment by petrochenkov** [comment]
@nikomatsakis @joshtriplett 
We a FCPing a modifier to the `#[macro_export]` attribute that changes the macro resolution behavior when `use_extern_macros` is enabled/stabilized and has no effect otherwise (including on older compilers, that's the critical requirement).

**Comment by scottmcm** [comment]
I'm still confused by what problem was actually agreed as needing solving in all this, but

@rfcbot reviewed

**Comment by joshtriplett** [comment]
On Mon, Jun 25, 2018 at 06:17:26PM -0700, scottmcm wrote:
> I'm still confused by what problem was actually agreed as needing solving in all this, but

With `use_extern_macros`, you can selectively import some but not all
macros from a crate. If you import a macro that wants to call other
macros from the same crate, it won't work with `use_extern_macros`.

A crate targeting current Rust can make its macros work with
`use_extern_macros` by making those macros use `$crate::macroname!` when
calling other macros from the same crate. However, some crates (such as
serde) still support versions of Rust that don't support the
`$crate::macroname!` syntax.

As a compatibility hack for those crates, they can put
`#[macro_export(local_inner_macros)]` on their macros, which will
automatically treat any macro calls within those macros as though they
used a `$crate::` prefix. And older versions of rustc will silently
ignore the argument to the `macro_export` attribute, allowing backward
compatibility.

**Comment by nikomatsakis** [comment]
@rfcbot resolved what-are-we-fcping

Thanks @dtolnay and @petrochenkov for the clarifications!

**Comment by nikomatsakis** [comment]
I checked @pnkfelix's box since they are not available just now.

**Comment by nikomatsakis** [comment]
@petrochenkov I think we can go ahead and r+ without waiting for the full FCP, if you want to rebase.

**Comment by nikomatsakis** [comment]
(As, again, this is not a final decision, and we could always revert)

**Comment by nikomatsakis** [comment]
(Oh, and because time is somewhat of the essence here.)

**Comment by sgrif** [comment]
Is there a writeup on the migration strategy for macro heavy crates, and how we can test with recent nightlies anywhere?

**Comment by joshtriplett** [comment]
@sgrif I don't know of a writeup to point to. The rough summary:

- If your crate doesn't need compatibility with old stable versions of Rust, then when calling any macro from within a macro, use the `$crate::` prefix.
- If your crate needs compatibility with old stable versions of Rust, then add the `#[macro_export(local_inner_macros)]` attribute to your macros, which will automatically act as though all inner macro calls have a `$crate::` prefix.
- If you have macros that *both* call helper macros *and* call macros you expect the caller to define, then you either need to use `$crate::` selectively on the helper macro calls or tell people they can't use `use` on your macros without also importing all your helper macros.

**Comment by petrochenkov** [comment]
>If you have macros that both call helper macros and call macros you expect the caller to define, then you either need to use $crate:: selectively on the helper macro calls or tell people they can't use use on your macros without also importing all your helper macros.

Or use `#[macro_export(local_inner_macros)]` and one more level of indirection for the caller-defined parts.

**Comment by joshtriplett** [comment]
@petrochenkov Thanks, I've edited my comment to take that possibility into account.

**Comment by sgrif** [comment]
Do macros which take other macros as callbacks need this as well?

**Pull request merged**